File: src/leetcode/p0100_0199/p0145_binary_tree_post_order_traversal/BinaryTreePostOrderTraversal.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 * 
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package leetcode.p0100_0199.p0145_binary_tree_post_order_traversal;

import common.provide.TreeNode;

import java.util.List;

public interface BinaryTreePostOrderTraversal {
    /**
     * Given the root of a binary tree, return the post order traversal
     * of its nodes' values.
     *
     * Follow up: Recursive solution ts trivial, could you do it
     * iteratively?
     *
     * @param root 0 <= root.size <= 100
     *             -100 <= Node.val <= 100
     * @return preorder travel sal
     */
    List<Integer> postorderTraversal(TreeNode root);
}

File: src/leetcode/p0100_0199/p0145_binary_tree_post_order_traversal/binary_tree_postorder_traversal.py
from abc import ABC, abstractmethod
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BinaryTreePostorderTraversal(ABC):
    """
    Given the root of a binary tree, return the postorder traversal of its nodes' values.
    """
    @abstractmethod
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        """
        Args:
            root:
                - The number of the nodes in the tree is in the range [0, 100].
                - -100 <= Node.val <= 100
        """
        pass

class BinaryTreePostorderTraversalRecursive(BinaryTreePostorderTraversal):
    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result: List[int] = []
        self.recursive(root, result);
        return result

    def recursive(self, root: Optional[TreeNode], result: List[int]) -> None:
        if not root:
            return;
        self.recursive(root.left, result)
        self.recursive(root.right, result)
        result.append(root.val)
